### PR TITLE
Add ATCMAC100 

### PR DESCRIPTION
**Pull request by maleehaakbarRS**
1. Add support of ATCDMAC100 for all types of :
   - Data Order 
   - Burst sizes  
   - Data width 
   - Modes of Transfer
   - DMA channels
   - Interrupts
2. EFR32MG12_LDMA is used as a reference point.
3. All DMA channels, addressing modes, transfer width and addressing 
    modes are tested on renode in parallel with the real prototype (asic01).  Link to all test cases 
    and their results are shared on the confluence page.
    https://rapidsilicon.atlassian.net/wiki/x/BIBGEg
4. Specifically, In Emulator/Main/Peripherals/DMA, two files DmaEngine.cs , and Request.cs are modified to support atcdmac100 implementation

### CODE REVIEW

File: src/Emulator/Main/Peripherals/Bus/Redirector.cs
Comment (jorge-rs) [inline src/Emulator/Main/Peripherals/Bus/Redirector.cs:1]: There is no real change to this file but a line eliminated. It should not appear as part of this PR.

File: src/Emulator/Main/Peripherals/Bus/SystemBus.cs
Comment (jorge-rs) [inline src/Emulator/Main/Peripherals/Bus/SystemBus.cs:1]: Also for this file: There is no real change to this file but a line eliminated. It should not appear as part of this PR.

File: src/Emulator/Main/Peripherals/Memory/MappedMemory.cs
Comment (jorge-rs) [inline src/Emulator/Main/Peripherals/Memory/MappedMemory.cs:1]: There is no real change to this file but a line eliminated. It should not appear as part of this PR.

File: src/Emulator/Peripherals/Peripherals/Miscellaneous/RS_SystemControlUnit.cs [hunk 1]
```diff
@@ -147,7 +147,7 @@ private void DefineRegisters()
             Registers.PufccControl.Define(this, 0x0)
                 .WithFlag(0, name:"pucc_rng_fre_out")
                 .WithFlag(1, name:"pucc_rng_fre_sel")
-                .WithFlag(1, name:"pucc_rng_fre_en"); // P2
+                .WithFlag(2, name:"pucc_rng_fre_en"); // P2
```
Comment (jorge-rs) [inline src/Emulator/Peripherals/Peripherals/Miscellaneous/RS_SystemControlUnit.cs:150]: Was this a bug you came across (meaning, that something was not working due to this? Or did you just came across this mistake?

File: src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs [hunk 2]
```diff
@@ -0,0 +1,578 @@
+//
+// Copyright (c) 2010-2023 Antmicro
```
Comment (jorge-rs) [inline src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs]: I guess it should be some RapidSilicon license rather that this one form Antmicro. We can check it from other code in EmbeddedCommon-dev

File: src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    SrcBurstSize:{blockSize},
+    SourceAddress: 0x{sourceAddress:X},
+    DestinationAddress: 0x{destinationAddress:X},
+    TranSize: {TranSize},
+    LinkAddress: 0x{(linkAddress << 2):X}
+}}";
+
+                
+// bits : starting position w.r.t register
+#pragma warning disable 649
```
Comment (jorge-rs) [inline src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs:527]: What is actually disabled with this and then re-enabled?

File: src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+                HalfWord = 0x1,
+                Word = 0x2,
+
+            }
+
+            [LeastSignificantByteFirst]
+            private struct Descriptor
+            {
+                public string PrettyString => $@"Descriptor {{        
+    Enable: {Enabled},
```
Comment (jorge-rs) [inline src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs:503]: This part of the code should be better indented. Funcionally speaking this doesn't matter but remember that to code is some sort of art, so it should look beautiful

File: src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+            private ulong LinkStructureAddress => (ulong)descriptor.linkAddress << 2;
+            private uint SourceIncrement => descriptor.SrcAddrCtrl == AddressMode.Fixed ? 0u : ((1u << (byte)descriptor.srcwidth));
+            private uint DestinationIncrement => descriptor.DstAddrCtrl == AddressMode.Fixed ? 0u : ((1u << (byte)descriptor.dstwidth));
+            private TransferType SizeAsTransferType => (TransferType)(1 << (byte)descriptor.srcwidth);
+            private int Bytes => (int)Math.Min(TranSize, BlockSizeMultiplier) << (byte)descriptor.srcwidth;
+            private bool SourceIncrementMode => descriptor.SrcAddrCtrl == AddressMode.Increment;
+            private bool SourceDecrementMode => descriptor.SrcAddrCtrl == AddressMode.Decrement;
+            private bool DestinationIncrementMode => descriptor.DstAddrCtrl == AddressMode.Increment;
+            private bool DestinationDecrementMode => descriptor.DstAddrCtrl == AddressMode.Decrement; 
+
```
Comment (jorge-rs) [inline src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs:462]: Additional empty line?

**Review by jorge-rs — CHANGES_REQUESTED** [review]
Please look at my comments and let's discuss about it

File: src/Emulator/Peripherals/Peripherals/Miscellaneous/RS_SystemControlUnit.cs (hunk 1 above)
Comment (maleehaakbarRS) [reply src/Emulator/Peripherals/Peripherals/Miscellaneous/RS_SystemControlUnit.cs:150]: Yes, this is a bug due to which rs_virgo.resc file failed to execute. So, correct it .

File: src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs (hunk 4 above)
Comment (maleehaakbarRS) [reply src/Emulator/Peripherals/Peripherals/DMA/ATCDMAC100.cs:503]: Adopt this code style from EFR32MG12_LDMA as that is used as a reference point.

**Review by jorge-rs — APPROVED** [review]

**Pull request merged**